dataRetriever/txpool/shardedTxPool.go

```diff
@@ -0,0 +1,254 @@
+package txpool
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/ElrondNetwork/elrond-go/data"
+	"github.com/ElrondNetwork/elrond-go/dataRetriever"
+	"github.com/ElrondNetwork/elrond-go/logger"
+	"github.com/ElrondNetwork/elrond-go/process"
+	"github.com/ElrondNetwork/elrond-go/storage"
+	"github.com/ElrondNetwork/elrond-go/storage/storageUnit"
+	"github.com/ElrondNetwork/elrond-go/storage/txcache"
+)
+
+var log = logger.GetOrCreate("dataretriever/txpool")
+
+// shardedTxPool holds transaction caches organised by destination shard
+type shardedTxPool struct {
+	mutex             sync.RWMutex
+	backingMap        map[string]*txPoolShard
+	mutexAddCallbacks sync.RWMutex
+	onAddCallbacks    []func(key []byte)
+	cacheConfig       storageUnit.CacheConfig
+	evictionConfig    txcache.EvictionConfig
+}
+
+type txPoolShard struct {
+	CacheID string
+	Cache   *txcache.TxCache
+}
+
+// NewShardedTxPool creates a new sharded tx pool
+// Implements "dataRetriever.TxPool"
+func NewShardedTxPool(config storageUnit.CacheConfig) (dataRetriever.ShardedDataCacherNotifier, error) {
+	err := verifyConfig(config)
+	if err != nil {
+		return nil, err
+	}
+
+	size := config.Size
+	evictionConfig := txcache.EvictionConfig{
+		Enabled:                         true,
+		CountThreshold:                  size,
+		ThresholdEvictSenders:           process.TxPoolThresholdEvictSenders,
+		NumOldestSendersToEvict:         process.TxPoolNumOldestSendersToEvict,
+		ALotOfTransactionsForASender:    process.TxPoolALotOfTransactionsForASender,
+		NumTxsToEvictForASenderWithALot: process.TxPoolNumTxsToEvictForASenderWithALot,
+	}
+
+	shardedTxPool := &shardedTxPool{
+		mutex:             sync.RWMutex{},
+		backingMap:        make(map[string]*txPoolShard),
+		mutexAddCallbacks: sync.RWMutex{},
+		onAddCallbacks:    make([]func(key []byte), 0),
+		cacheConfig:       config,
+		evictionConfig:    evictionConfig,
+	}
+
+	return shardedTxPool, nil
+}
+
+func verifyConfig(config storageUnit.CacheConfig) error {
+	if config.Size < 1 {
+		return fmt.Errorf("size must be a positive number")
```

Comment:
declare errors for this.

Reply:
Declared them in `dataRetriever/errors.go`.